core/src/main/java/com/linecorp/armeria/client/DefaultEventLoopScheduler.java

```diff
@@ -0,0 +1,388 @@
+                            .collect(toImmutableList());
+        checkArgument(maxEventLoopPerEndpoint <= eventLoops.size(),
+                      "maxEventLoopPerEndpoint: %s (expected: <= %s)", maxEventLoopPerEndpoint,
+                      eventLoops.size());
+        if (maxEventLoopPerEndpoint == 0) {
+            this.maxEventLoopPerEndpoint = eventLoops.size();
+        } else {
+            this.maxEventLoopPerEndpoint = maxEventLoopPerEndpoint;
+        }
+
+        for (Map.Entry<String, Integer> entry : maxEventLoops.entrySet()) {
+            checkArgument(entry.getValue() <= eventLoops.size(),
+                          "maxEventLoopPerEndpoint for %s: %s (expected: <= %s)",
+                          entry.getKey(), entry.getValue(), eventLoops.size());
+        }
+        this.maxEventLoops = ImmutableMap.copyOf(maxEventLoops);
+    }
+
+    @Override
+    public ReleasableHolder<EventLoop> acquire(Endpoint endpoint) {
+        requireNonNull(endpoint, "endpoint");
+        final State state = state(endpoint);
+        final Entry acquired = state.acquire();
+        cleanup();
+        return acquired;
+    }
+
+    @VisibleForTesting
+    List<Entry> entries(Endpoint endpoint) {
+        return state(endpoint).entries();
+    }
+
+    private State state(Endpoint endpoint) {
+        final String authority = endpoint.authority();
+        final State state = states.get(authority);
+        if (state != null) {
+            return state;
+        }
+
+        final Integer maxEventLoop = maxEventLoops.get(authority);
+        return states.computeIfAbsent(
+                authority, unused -> new State(eventLoops,
+                                               maxEventLoop != null ? maxEventLoop : maxEventLoopPerEndpoint,
+                                               currentIndex));
+    }
+
+    /**
+     * Cleans up empty entries with no activity for more than 1 minute. For reduced overhead, we perform this
+     * only when 1) the last clean-up was more than 1 minute ago and 2) the number of acquisitions % 256 is 0.
+     */
+    private void cleanup() {
+        if ((++counter & 0xFF) != 0) { // (++counter % 256) != 0
+            return;
+        }
+
+        final long currentTimeNanos = System.nanoTime();
+        if (currentTimeNanos - lastCleanupTimeNanos < CLEANUP_INTERVAL_NANOS) {
+            return;
+        }
+
+        for (final Iterator<State> i = states.values().iterator(); i.hasNext();) {
+            final State state = i.next();
+            final boolean remove;
+
+            synchronized (state) {
+                remove = state.allActiveRequests == 0 &&
+                         currentTimeNanos - state.lastActivityTimeNanos >= CLEANUP_INTERVAL_NANOS;
+            }
+
+            if (remove) {
+                i.remove();
+            }
+        }
+
+        lastCleanupTimeNanos = System.nanoTime();
+    }
+
+    private static final class State {
```

Comment:
Could we introduce some type hierarchy for this class and provide a special implementation that performs linear search for the case of `maxNumEventLoops <= N` where N is fairly small? A little bit of benchmarking may be necessary to determine N. Consider taking care of this in a follow-up PR. Originally suggested by @anuraaga some time ago.

Reply:
Will do it in the next PR. (or the next of the next PR 🤣)